File: src/main/java/club/mindtech/mindbot/commands/Commands.java
package club.mindtech.mindbot.commands;

import club.mindtech.mindbot.commands.help.CommandHelp;
import club.mindtech.mindbot.commands.ping.CommandPing;
import club.mindtech.mindbot.commands.poll.CommandPoll;
import net.dv8tion.jda.api.interactions.commands.build.CommandData;

import java.util.Map;
import java.util.TreeMap;

public class Commands {

    private static final Map<String, BaseCommand> COMMANDS = new TreeMap<>();

    public static void register(BaseCommand command) {
        COMMANDS.put(command.getName(), command);
    }

    private static void registerCommands() {
        register(new CommandPing());
        register(new CommandPoll());
        register(new CommandHelp());
    }

    public static CommandData[] getSlashCommandData() {
        registerCommands();
        return COMMANDS
                .values()
                .stream()
                .map(BaseCommand::getCommandData)
                .toArray(CommandData[]::new);
    }

    public static boolean isCommand(String command) {
        return COMMANDS.containsKey(command);
    }

    public static BaseCommand getCommand(String command) {
        return COMMANDS.get(command);
    }
}

File: src/main/kotlin/club/mindtech/mindbot/commands/Commands.kt
package club.mindtech.mindbot.commands

import club.mindtech.mindbot.commands.archive.CommandArchive
import club.mindtech.mindbot.commands.help.CommandHelp
import club.mindtech.mindbot.commands.ping.CommandPing
import club.mindtech.mindbot.commands.poll.CommandPoll
import net.dv8tion.jda.api.interactions.commands.build.CommandData

private val COMMAND_MAP = setOf(
    CommandPing(),
    CommandHelp(),
    CommandPoll(),
    CommandArchive()
).associateBy { it.name }

val registeredCommands = COMMAND_MAP.toMap() // to make a copy

fun getSlashCommandData(): Array<CommandData> {
    return COMMAND_MAP.values.map { it.getCommandData() }.toTypedArray()
}

fun getCommand(command: String?): BaseCommand? {
    return if (command == null) {
        null
    } else {
        COMMAND_MAP[command]
    }
}
